tests/RbacE2ETest.php

```diff
@@ -0,0 +1,409 @@
+                ],
+                []
+            );
+        }
+
+        $biscuit = $builder->build($this->rootKeyPair->private());
+
+        return $biscuit->toBase64();
+    }
+
+    /**
+     * Create a token with resource-specific roles
+     */
+    private function createUserTokenWithResourceRoles(string $userId, array $resourceRoles): string
+    {
+        $builder = new BiscuitBuilder();
+
+        $builder->addCodeWithParams('user({id})', ['id' => $userId], []);
+
+        foreach ($resourceRoles as $resource => $roles) {
+            $builder->addCodeWithParams(
+                'user_roles({id}, "api", {resource}, {roles})',
+                [
+                    'id' => $userId,
+                    'resource' => $resource,
+                    'roles' => $roles,
+                ],
+                []
+            );
+        }
+
+        $biscuit = $builder->build($this->rootKeyPair->private());
+
+        return $biscuit->toBase64();
+    }
+
+    /**
+     * Attenuate token to read-only access
+     */
+    private function attenuateToReadOnly(string $tokenBase64): string
+    {
+        $biscuit = Biscuit::fromBase64($tokenBase64, $this->rootKeyPair->public());
+
+        // Add a block that restricts operations to read-only
+        $block = new \Biscuit\Auth\BlockBuilder();
+        $block->addCode('check if operation("read")');
+
+        $attenuatedBiscuit = $biscuit->append($block);
+
+        return $attenuatedBiscuit->toBase64();
+    }
+
+    /**
+     * Authorize a user operation on a resource using RBAC rules
+     *
+     * This implements the RBAC pattern from the Biscuit recipe:
+     * https://www.biscuitsec.org/docs/recipes/role-based-access-control/
+     */
+    private function authorize(
+        string $tokenBase64,
+        string $userId,
+        string $operation,
+        string $resource
+    ): bool {
+        try {
+            // Parse and verify token
+            $biscuit = Biscuit::fromBase64($tokenBase64, $this->rootKeyPair->public());
+
+            // Build authorizer with RBAC rules
+            $authBuilder = new AuthorizerBuilder();
+
+            // Add role definitions for the requested resource
+            // (role definitions map permissions for each resource type)
+            $this->addRoleDefinitions($authBuilder, $resource);
+
+            // Add RBAC authorization logic (from Biscuit recipe)
+            $this->addRbacRules($authBuilder);
+
+            // Add facts for current operation
+            $authBuilder->addCodeWithParams(
+                'operation({op}); resource({res});',
+                [
+                    'op' => $operation,
+                    'res' => $resource,
+                ],
+                []
+            );
+
+            // Build authorizer and check authorization
+            $authorizer = $authBuilder->build($biscuit);
+            $policy = $authorizer->authorize();
+
+            // Policy 0 = allow
+            return $policy === 0;
```

Comment:
it’s a bit redundant imo, since you control the policies, and assuming that `Authorizer::authorize()` raises an exception in case of error